ir/pattern/mod.rs

```diff
@@ -348,3 +349,84 @@ impl fmt::Display for ValueType {
         }
     }
 }
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+enum VariableDependencyMode {
+    Required,
+    Producing,
+    Referencing,
+}
+
+impl BitAndAssign for VariableDependencyMode {
+    fn bitand_assign(&mut self, rhs: Self) {
+        match (*self, rhs) {
+            (Self::Producing, _) | (_, Self::Producing) => *self = Self::Producing,
+            (Self::Required, _) | (_, Self::Required) => *self = Self::Required,
+            (Self::Referencing, Self::Referencing) => (),
+        }
+    }
+}
+
+impl BitOrAssign for VariableDependencyMode {
+    fn bitor_assign(&mut self, rhs: Self) {
+        match (*self, rhs) {
+            (Self::Required, _) | (_, Self::Required) => *self = Self::Required,
+            (Self::Referencing, _) | (_, Self::Referencing) => *self = Self::Referencing,
+            (Self::Producing, Self::Producing) => (),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct VariableDependency<'a> {
+    mode: VariableDependencyMode,
+    places: Vec<&'a Constraint<Variable>>,
```

Comment:
Do we want to call places `references`?
Just to clarify: It's all places that a variable is referenced regardless of what the mode there was?

Reply:
Yeah let's use `referencing_constraints`